src/meta/src/manager/env.rs

```diff
@@ -304,87 +347,89 @@ impl MetaSrvEnv {
-            }
-            None => None,
-        };
+                );
 
-        let event_log_manager = Arc::new(start_event_log_manager(
-            opts.event_log_enabled,
-            opts.event_log_channel_max_size,
-        ));
-        let hummock_seq = meta_store_sql
-            .clone()
-            .map(|m| Arc::new(SequenceGenerator::new(m.conn)));
-        let sql_id_gen_manager = if let Some(store) = &meta_store_sql {
-            Some(Arc::new(SqlIdGeneratorManager::new(&store.conn).await?))
-        } else {
-            None
+                Self {
+                    id_gen_manager_impl: IdGenManagerImpl::Sql(Arc::new(
+                        SqlIdGeneratorManager::new(&sql_meta_store.conn).await?,
+                    )),
+                    system_param_manager_impl: SystemParamsManagerImpl::Sql(
+                        system_param_controller,
+                    ),
+                    meta_store_impl: meta_store_impl.clone(),
+                    notification_manager,
+                    stream_client_pool,
+                    idle_manager,
+                    event_log_manager,
+                    cluster_id,
+                    hummock_seq: Some(Arc::new(SequenceGenerator::new(
+                        sql_meta_store.conn.clone(),
+                    ))),
+                    opts: opts.into(),
+                }
+            }
         };
+        Ok(env)
+    }
 
-        Ok(Self {
-            id_gen_manager,
-            sql_id_gen_manager,
-            meta_store,
-            meta_store_sql,
-            notification_manager,
-            stream_client_pool,
-            idle_manager,
-            event_log_manager,
-            system_params_manager,
-            system_params_controller,
-            cluster_id: cluster_id.unwrap(),
-            opts: opts.into(),
-            hummock_seq,
-        })
+    pub fn kv_meta_store_ref_checked(&self) -> MetaStoreRef {
+        self.meta_store_impl.assert_kv_ref().clone()
     }
 
-    pub fn meta_store_ref(&self) -> MetaStoreRef {
-        self.meta_store.clone().unwrap()
+    pub fn kv_meta_store_checked(&self) -> &MetaStoreRef {
+        self.meta_store_impl.assert_kv_ref()
     }
 
-    pub fn meta_store_checked(&self) -> &MetaStoreRef {
-        self.meta_store.as_ref().unwrap()
+    pub fn sql_meta_store_ref_checked(&self) -> SqlMetaStore {
+        self.meta_store_impl.assert_sql_ref().clone()
     }
 
-    pub fn meta_store(&self) -> Option<&MetaStoreRef> {
-        self.meta_store.as_ref()
+    pub fn meta_store_ref(&self) -> MetaStoreImpl {
+        self.meta_store_impl.clone()
     }
 
-    pub fn sql_meta_store(&self) -> Option<SqlMetaStore> {
-        self.meta_store_sql.clone()
+    pub fn meta_store(&self) -> &MetaStoreImpl {
+        &self.meta_store_impl
     }
 
-    pub fn id_gen_manager(&self) -> &IdGeneratorManager {
-        self.id_gen_manager.as_ref().unwrap()
+    pub fn kv_id_gen_manager(&self) -> &IdGeneratorManager {
```

Comment:
What about not exposing such methods but letting caller write `env.id_gen_manager().as_kv()`? Same for meta store.